### ISSUE TITLE
Creating TensorFlow device issue and perfromance drop

### ISSUE THREAD

**Issue by civilmanxx** [tensorflow/tensorflow#5480]
Now I have 4 GeForce GTX TITAN X on my server. I want to run 4 tensor flow app independently on different GPU. I use the following code to assign GPU for different app, such as the 1st app use gpu:0 and 2nd app use gpu:1.

**tf.device('/gpu:0')**

However, whenever i run any app individually or all apps at same time, each app CMD window still outputs as below. Each creates 4 devices no matter other one(s) use these device or not.

After starting 3 apps, which are assigned the first 3 GPU (0,1,2) for each of them, then i start the 4th app for GPU:3, which is the 4th GPU, the app has dramatic performance drop.

Now i am using **Ubuntu 16.04, cuda 8 + cuDNN 5.1.** I do not know why this happen. I am not clear about if each app could run separately on different GPU on the same machine. 

**I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:0a:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX TITAN X, pci bus id: 0000:09:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:2) -> (device: 2, name: GeForce GTX TITAN X, pci bus id: 0000:06:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:3) -> (device: 3, name: GeForce GTX TITAN X, pci bus id: 0000:05:00.0)**

**Comment by civilmanxx** [comment]
also, i meet more than once that start a new app on different GPU make other app (process) on different GPU out of memory.

**Comment by sherrym** [comment]
Hi @civilmanxx ,

All these GPUs share the same CPU and memory. Could you please provide the output pf top, htop, nvidia-smi, vmstat while running 1, 2, 3 and 4 apps and see what the difference might be? Thanks.

Sherry

**Comment by civilmanxx** [comment]
let me clarify, out of memory at here is CUDA device out of memory. You mean all apps share GPU memory? each GeForce GTX TITAN X has 12GB GPU memory. Once an app start, it occupies all 4 GPU memory? 12GB * 4 = 48 GB. But my app is very small, just run Inception V1 for single image feature extraction on different GPU. 4 Apps are the same.

**Comment by civilmanxx** [comment]
I just use "Nvidia X Server Settings" tool to monitor GPU memory usage. I only start one app but all GPU memory on 4 GPUs are occupied. Is this make sense? If start only one then occupy all GPU, then **tf.device('/gpu:0')** become useless.

**Comment by civilmanxx** [comment]
Now i use the following code to let memory usage growth on demand but not allocate all at once.

**config=tf.ConfigProto()
config.gpu_options.allow_growth=True
sess=tf.Session(config=config)**

I run one app only and assign current app to use GPU:3 (the 4th one) by the code: **tf.device('/gpu:3'):**

However, when i monitor the memory usage by "Nvidia X Server Settings" tool. GPU:3 uses 111M memory, which should be. But both GPU:1 and GPU:2 uses 111M as well, which mean even if I do not assign my app to use these GPUs and set config **growth=True**. In fact GPU:0 use 111M as well for my app.

So this mean when we set GPU memory config to growth mode, it growths on all devices at same time. but not growth one after another, even if I assign which GPU to use. As i have 4 GPUs and i only run one app on one GPU, Tenor flow waste 3 copies of GPU memory.

**Comment by civilmanxx** [comment]
So I just want to know if tensor flow support real separation on different GPU for different process.
For example, one process is assigned to use GPU:0, so when the session start to run, tensor flow only create one device gpu:0 for use and all the rest GPUs are in the idle status. This is make sense, in every aspect.

**Comment by flx42** [comment]
@civilmanxx: for weak isolation, use environment variable [CUDA_VISIBLE_DEVICES](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#env-vars).
For real isolation, you might want to run TensorFlow in [Docker](https://www.tensorflow.org/versions/r0.11/get_started/os_setup.html#docker-installation), when using [nvidia-docker](https://github.com/NVIDIA/nvidia-docker) you can set a variable `NV_GPU` to isolate devices.

**Comment by civilmanxx** [comment]
I found there should be bug there. Now my single process app is assigned to use gpu:3,  but when i monitor the GPU utilization in  "Nvidia X Server Settings" tool,  gpu:3 utilization is 0%, but gpu:0 utilization is 5%. That's to say  **tf.device('/gpu:3'):** is not in effect at all.

**Comment by civilmanxx** [comment]
@flx42 Use **virtual** machine Docker to isolate **physical** hardware GPU... ...

Well, this is an interesting idea. I feel this mechanism is not very humanize. It is a detour. Tensor flow GPU version is build on CUDA. I think it is CUDA framework to decide how GPUs should be used, based on pass in parameter, like device ID assignment. I am not sure this is CUDA's issue or tensor flow's issue. Isolation should be a very fundamental job need to do. Is there any plan to support this natively? I do not know it belongs to whose job, NVidia or Google.

**Comment by flx42** [comment]
@civilmanxx Docker is not a virtual machine, applications will show up as regular processes on your machine.

**Comment by asimshankar** [comment]
@civilmanxx : Similar to what was described in #5066 : TensorFlow will allocate all the memory across all the GPUs accessible to it. If you want to run independent TensorFlow programs, say one on each GPU, then as @flx42 mentioned, you probably want to do something like this:

```sh
# Run myprogram.py on GPU 0
export CUDA_VISIBLE_DEVICES="0" 
myprogram.py
# And another instance on GPU1
export CUDA_VISIBLE_DEVICES="1"
myprogram.py
# And another instance on GPU2
export CUDA_VISIBLE_DEVICES="2"
```

Note that the `CUDA_VISIBLE_DEVICES` environment variable restricts what GPUs the TensorFlow runtime "sees", and in the setup above, you want to refer to the GPU as `gpu:0` in all the instances.

Let us know if that works out for you, or if I misunderstood.

**Comment by civilmanxx** [comment]
I set export CUDA_VISIBLE_DEVICES="1", but it uses gpu:2 instead. The index is 0,1,2,3.

**Comment by civilmanxx** [comment]
after i start the 1st app, which is using gpu:2 (though i set it to 1) and then set device=3 for the 2nd app in another process. The 2nd app use gpu:0 instead.

**Comment by asimshankar** [comment]
@civilmanxx : Could you elaborate on what you mean by "it uses gpu:2 instead"? The device names in TensorFlow will correspond to the GPUs accessible to it, while the `CUDA_VISIBLE_DEVICES` environment variable is what the CUDA framework uses to restrict visibility. So, if `CUDA_VISIBLE_DEVICES="1"`, then TensorFlow programs should not be able to see multiple GPUs.

Perhaps you can provide a reproducible example of the program you're using and the output of things like `nvidia-smi` and others like @sherrym mentioned above, to help explain what the problem is?

**Comment by flx42** [comment]
Maybe @civilmanxx is confused with the output of `nvidia-smi`. Be aware that the device ordering provided by CUDA might be different from the device ordering shown by `nvidia-smi`.
Using `CUDA_VISIBLE_DEVICES=0` might actually give you the **latest** device when checking utilization with `nvidia-smi`.
To get consistent numbering, use `CUDA_​DEVICE_​ORDER=PCI_BUS_ID`, this is documented [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#env-vars)

**Comment by civilmanxx** [comment]
I am using Ubuntu 16.04. If go to "Nvidia X server settings" you can see the list of GPUs. It shows each GPU utilization and used memory. I suppose the CUDA_VISIBLE_DEVICES="index", at here index 0,1,2,3 should be same as the index show in "Nvidia X server settings". is that correct?

**Comment by civilmanxx** [comment]
if base on the post from @flx42 , they are different.

**Comment by civilmanxx** [comment]
@flx42 you mean i need run export cmd twice as below, is that right?

export CUDA_​DEVICE_​ORDER = PCI_BUS_ID 
export CUDA_VISIBLE_DEVICES="0"

**Comment by flx42** [comment]
@civilmanxx: correct, but the `CUDA_DEVICE_ORDER` export needs to be done only once. Whereas you export `CUDA_VISIBLE_DEVICES` before each run.

**Comment by civilmanxx** [comment]
i get error when i run **export CUDA_​DEVICE_​ORDER = PCI_BUS_ID** 

`CUDA_​DEVICE_​ORDER=PCI_BUS_ID': not a valid identifier

**Comment by civilmanxx** [comment]
I am using cuda 8.0

**Comment by flx42** [comment]
Come on, that's obviously not a CUDA problem if you can't `export` a variable. You just messed up your encoding.

**Comment by civilmanxx** [comment]
but i can run **export** **CUDA_VISIBLE_DEVICES="0"**  with no issue. The cmd **export** **CUDA_​DEVICE_​ORDER=PCI_BUS_ID** is right.

**Comment by civilmanxx** [comment]
I am not clear about why copy paste cmd from browser to terminal does not work. type cmd in terminal directly works.

**Comment by civilmanxx** [comment]
It turns out, even if i run these 2 export cmds before start my app. the assigned gpu index in export cmd is still different from the true running one. I set gpu:3 but gpu:1 is running...

**Comment by civilmanxx** [comment]
and I run another app in the 2nd process and set gpu:3 again, but it uses gpu:0.

**Comment by vrv** [comment]
One point of clarification.  "/gpu:0" will always be the first logical GPU available in the process.

So if you set CUDA_VISIBLE_DEVICES=2, your program should still use "/gpu:0", because only one physical GPU is made visible to the process as "logical GPU id 0".

If you set CUDA_VISIBLE_DEVICES=4,2, then "/gpu:0 would be the 4th physical GPU, and "/gpu:1" would be the 2nd physical GPU.

**Comment by civilmanxx** [comment]
I do not set /gpu:0 or /gpu:3 in side of the python code at this time. I only use CUDA_VISIBLE_DEVICES="3" to set current process see one GPU only. As I already set CUDA_​DEVICE_​ORDER=PCI_BUS_ID, I expect that current process only see the visible device, which is 3. But in fact it is not.

**Comment by civilmanxx** [comment]
in other word, **CUDA_​DEVICE_​ORDER=PCI_BUS_ID** is not effective.

**Comment by civilmanxx** [comment]
the code i use is **tensorflow/tensorflow/models/slim/train_image_classifier.py** There is no device assignment code in the file.

**Comment by girving** [comment]
@asimshankar Is more information necessary to debug this?  From my understanding of the thread so far, @civilmanxx says that `CUDA_VISIBLE_DEVICES` doesn't work, but we have no idea why.

**Comment by asimshankar** [comment]
I'm not quite sure I still fully understand the problem here.

This issue began with a question around being able to use one GPU each in 4 simultaneously running programs. And the code being used is [train_image_classifier.py](https://github.com/tensorflow/models/blob/master/slim/train_image_classifier.py).

The suggestion was to use `CUDA_VISIBLE_DEVICES` to restrict visibility of each instance of the program to a single GPU.

What isn't clear to me is what happened after this. @civilmanxx : Is it that after setting `CUDA_VISIBILE_DEVICES`, each instance of the program still uses multiple GPUs or is it that you can't stabilize on _which_ one of the GPUs is being used?

It will be really helpful if you can provide detailed instructions on reproducing the problem - like what steps you're running, which release of TensorFlow etc. (all the details asked for in the new issue template).

**Comment by civilmanxx** [comment]
The thread 1 use **tf.device('/gpu:0')** in python code to assign GPU.  But to simplify the issue for easier investigation,  just try **https://github.com/tensorflow/models/blob/master/slim/train_image_classifier.py** is enough.

run these 2 cmd:

export CUDA_​DEVICE_​ORDER = PCI_BUS_ID 
export CUDA_VISIBLE_DEVICES=3 

and then run **train_image_classifier.py**, you will notice that the true running GPU is not (PCI) 3.

**Comment by asimshankar** [comment]
But it is running on a single GPU?
If so, then this is perhaps an issue you want to investigate with the CUDA library folks as my understanding is that TensorFlow is simply working with the single GPU device that CUDA has made accessible to it.

**Comment by civilmanxx** [comment]
I do not want to spend more time on this now. Now i already do not care which is which.

**Comment by ugtony** [comment]
I ran the program as @civilman628 did
```
with tf.device("/gpu:2"):
   ...

sess = tf.Session(config=tf.ConfigProto(gpu_options=tf.GPUOptions(allow_growth = True), 
    allow_soft_placement = True))     
```
The program allocated the memory on GPU-2 as I expected using Tensorflow 0.11, 
but allocated the memory on all GPUs when I moved the project to Tensorflow 0.12.1.

I wonder if it is a bug. Since the behavior is inconsistent.
